columnflow/production/cms/seeds.py

```diff
@@ -23,12 +18,15 @@
     uses={
         # global columns for event seed
         "run", "luminosityBlock", "event",
-        optional("Photon.pt"), optional("SV.pt"),
-        # first-object columns for event seed
-        optional("Tau.jetIdx"), optional("Tau.decayMode"),
-        optional("Muon.jetIdx"), optional("Muon.nStations"),
-        optional("Jet.nConstituents"), optional("Jet.nElectrons"), optional("Jet.nMuons"),
+        "Pileup.nPU",
+        # columns needed to extract object counts only
+        "Photon.pt", "SV.pt", "FatJet.pt", "SubJet.pt",
         optional("GenJet.pt"), optional("GenPart.pt"),
+        # per-object columns for further hashing
+        "Electron.jetIdx", "Electron.seediPhiOriY",
+        "Muon.jetIdx", "Muon.nStations",
+        "Tau.jetIdx", "Tau.decayMode",
+        "Jet.nConstituents", "Jet.nElectrons", "Jet.nMuons",
```

Comment:
Why is the `optional` flag being dropped here? Unfortunately, this will definitely introduce breaking changes for analyses that rely on running `determinisic_seeds` later in the workflow, e.g. during `ProduceColumns`, since in that case not all columns defined in the original NanoAOD would be available.

I think we should keep this implementation as flexible as possible and not impose any hard requirements on input columns. Can this be adapted to use the `optional` flag for all columns except the mandatory `run`, `luminosityBlock` and `event`?

Reply:
Yes, we can make them optional again. The motivation for dropping it was to be sensitive the possible broken inputs where certain columns should exist, but are in fact missing, leading to different seeds without knowing.

But in order not to break existing code, we could introduce a `missing_column_behavior`, that is `"ignore"` by default (old behavior), but can also be set to `"warning"` or `"raise"`.

Reply:
Right, that's also a valid point. The `missing_column_behavior` sounds like a reasonable way to make this more flexible, but I'm thinking whether we should implement that separately or extend the existing mechanisms to make sure unexpectedly missing columns don't go unnoticed.

We could in principle do this via the `check_used_columns` mechanism in the base class. At the moment that only supports either ignoring the columns or raising an exception, depending on whether the column is wrapped by the `optional` modifier. 

But maybe we can refine this behavior to issue a `warning` if certain optional columns are missing? This could be controlled by another aux flag à la `skip` or `optional`. And as a user interface we could add an `optional_warn` column wrapper that sets the flag. What do you think, @riga?

At any rate, the implementation would still need to check if these columns actually exist before using them to calculate the seeds.

Reply:
I like the idea of having a variant of the `optional` modifier that issues a warning in case columns are not existing. With that, all three use cases would be covered.

I would prepare this for another PR though if you agree.

> At any rate, the implementation would still need to check if these columns actually exist before using them to calculate the seeds.

In *this* PR, we should do exactly that imho. Would you be fine if we just default to printing warnings when columns are missing?

Reply:
In 406930d0, we made the columns optional again and added the warnings :)

Reply:
OK, great! Yes, let's do the `optional` variant in another PR. For this PR it would be fine to just ignore the columns and directly print the warnings, as you suggest.